### PR TITLE
Jetpack: add custom validity for input type URL

### LINKED ISSUE #27119: Form Block: Website child block -> the error message for "entering URL" needs to be more precise (->https:// is expected)
Current "error-help message":

Image

Suggestion:

-> Please enter a full URL - https://www.example.com

### PR DESCRIPTION
**Pull request by CGastrell**
Use browser's native support for HTML input element validation with custom text. Inputs type URL never specify any expectation on the value, yet the URL is considered invalid unless it includes protocol (http/https).

Fixes #27119 

Addresses #13914

#### Changes proposed in this Pull Request:
- use `setCustomValidity` to change the hint text shown when trying to submit a form and a required URL input is not filled correctly

Before:
image

After:
image

Note: this change is a "fix", but native support will override the value once the input is changed. This works as a better hinting for the user, further customization would require a more complex solution

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
Insert a form and add a "Website" input (lately renamed to "URL Field"). Try submitting the form with different values on the website field.

### CODE REVIEW

**Review by digitalwaveride — CHANGES_REQUESTED** [review]
Thank you Christian!
Not sure if related to this PR but there is something off on the submission view -> `"`:

image


A little other thing I noticed:
When typing a url without any "prefix" ... like www.somesite.com I get the error via the "tooltip" 
and it looks like this:
image

The moment though I start to type something like this "http:/" or "ht:/"
image

it seems to try to submit, and shows an error like this:

image


Regarding the copy of the error text:

image

I personally find it a bit hard to read, and in these cases like to give users an example that they can mimic:

`Please enter a valid URL - https://www.example.com`

**Review by digitalwaveride — DISMISSED** [review]
Perfect! Works as advertised!
Thanks Christian!!! 🚀

File: projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php [hunk 1]
```diff
@@ -4030,6 +4030,15 @@ public function validate() {
 		}
 
 		switch ( $field_type ) {
+			case 'url':
+				if ( ! is_string( $field_value ) || empty( $field_value ) || ! preg_match(
+					'%^(?:(?:https?|ftp)://)?(?:\S+(?::\S*)?@|\d{1,3}(?:\.\d{1,3}){3}|(?:(?:[a-z\d\x{00a1}-\x{ffff}]+-?)*[a-z\d\x{00a1}-\x{ffff}]+)(?:\.(?:[a-z\d\x{00a1}-\x{ffff}]+-?)*[a-z\d\x{00a1}-\x{ffff}]+)*(?:\.[a-z\x{00a1}-\x{ffff}]{2,6}))(?::\d+)?(?:[^\s]*)?$%iu',
```
Comment (samiff) [inline projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php:4057]: For my own curiosity, was there an external source for this regex?

**Review by samiff — DISMISSED** [review]
It's a shame that there isn't a simple function like `is_email` to leverage 🙂

A couple thoughts that come to my mind here:

1. If we do go the regex route, perhaps we can abstract that functionality to a shared helper function somewhere as it'd seem quite useful to have around (and tested). I tried looking to see if we already did that somewhere, though the closest I found was using [FILTER_VALIDATE_URL](https://www.php.net/manual/en/filter.filters.validate.php) which isn't perfect as it will not work for internationalized domain names (IDNs)
2. This PR as-is won't work for IDNs, example: https://παράδειγμα.test
3. Is there a great reason to validate the URL field type to begin with? As a user, entering `example.com` and then having to enter `http://example.com` to be able to submit the form feels nagging. Does the validation to URL spec provide much value to the website owner I wonder?

**Comment by CGastrell** [comment]
@samiff completely agree with you, yet I’d stick to a more constrained
scope:

- improve the default validation message browsers offer for the `type=url`
inputs, as it’s not useful enough
- we’re on a long plan to improve overall UX regarding forms/inputs, and
we’ll surely take over all validations eventually (as opposed to rely on
browsers’ default behavior)


I’ll get back at the source of the URL regex, I’m sure it came from SO, but thanks for reminding me of quoting this kind of things.

EDIT: regex came from https://urlregex.com/ but, TBH, it's just a backend check to better support what the UI is already trying to take care of. I'll use the simpler `FILTER_VALIDATE_URL` approach.

**Comment by CGastrell** [comment]
@samiff 

I've created #27396 to keep track of your comment regarding an improvement for IDNS support both at code and UX levels.

Meanwhile, I've updated the PR so it allows for more flexible URLs (still no IDNS support), using `type=text` on the frontend but parsing as `type=url` on backend with custom regex validation (allowing URLs without protocol)

**Review by samiff — APPROVED** [review]
I'm a tiny bit hesitant on the loss of IDN support for the URL field with this PR, but approving since we have #27396 open and it sounds like there are larger plans for validation forthcoming.

**Pull request merged**